- case: managed_with_reader_future_result_callback
  disable_cache: true
  main: |
    from returns.context import ReaderFutureResult
    from returns.pipeline import managed
    from returns.result import Result

    def use(acquired_value: int) -> ReaderFutureResult[float, str, bool]:
        ...

    def release(
        acquired_value: int,
        use_value: Result[float, str],
    ) -> ReaderFutureResult[None, str, bool]:
        ...

    reveal_type(managed(use, release))  # N: Revealed type is 'def (returns.context.requires_context_future_result.RequiresContextFutureResult[builtins.int*, builtins.str*, builtins.bool*]) -> returns.context.requires_context_future_result.RequiresContextFutureResult[builtins.float*, builtins.str*, builtins.bool*]'


- case: managed_with_reader_future_result_error
  disable_cache: true
  main: |
    from returns.context import ReaderFutureResult
    from returns.pipeline import managed
    from returns.result import Result

    def use(acquired_value: int) -> ReaderFutureResult[float, str, bool]:
        ...

    def release(
        acquired_value: int,
        use_value: Result[float, int],
    ) -> ReaderFutureResult[int, str, str]:
        ...

    reveal_type(managed(use, release))  # TODO: this type is really strange!
  out: |
    main:14: error: Argument 1 to "_managed" has incompatible type "Callable[[int], RequiresContextFutureResult[float, str, bool]]"; expected "Callable[[int], IOResult[float, int]]"
    main:14: error: Argument 2 to "_managed" has incompatible type "Callable[[int, Result[float, int]], RequiresContextFutureResult[int, str, str]]"; expected "Callable[[int, Result[float, int]], IOResult[None, int]]"
    main:14: note: Revealed type is 'def (returns.io.IOResult[builtins.int*, builtins.int*]) -> returns.io.IOResult[builtins.float*, builtins.int*]'
